### PR TITLE
Histograms: Expose and fix #12552

### LINKED ISSUE #12552: histograms: scraping classic and native histograms in parallel misses some native histograms
### What did you do?

- Expose two or more native histograms within the same metric family (i.e. with the same name) on the same target, both with classic buckets, too.
- Configure `scrape_classic_histograms: true` in the scrape config.
- Run Prometheus with `--enable-feature=native-histograms`.

### What did you expect to see?

Ingest the classic and the native histograms for both histograms in the metric family.

### What did you see instead? Under which circumstances?

Both classic histograms are ingested, only one native.

Note that two histograms with different names are ingested fine.

### System information

_No response_

### Prometheus version

```text
rometheus, version  (branch: , revision: 162612ea8627447be0b8fa170268b76524124bbc)
  build user:       
  build date:       
  go version:       go1.20.4
  platform:         linux/amd64
  tags:             unknown
```

### Prometheus configuration file

_No response_

### Alertmanager version

_No response_

### Alertmanager configuration file

_No response_

### Logs

_No response_

### PR DESCRIPTION
**Pull request by beorn7**

### CODE REVIEW

**Comment by beorn7** [comment]
@krajorama please have a look.

File: model/textparse/protobufparse.go [hunk 1]
```diff
@@ -411,6 +411,14 @@ func (p *ProtobufParser) Next() (Entry, error) {
 			p.metricPos++
 			p.fieldPos = -2
 			p.fieldsDone = false
+			// If this is a metric family containing native
+			// histograms, we have to switch back to native
+			// histograms after parsing a classic histogram.
+			if p.state == EntrySeries &&
+				(t == dto.MetricType_HISTOGRAM || t == dto.MetricType_GAUGE_HISTOGRAM) &&
+				isNativeHistogram(p.mf.GetMetric()[0].GetHistogram()) {
```
Comment (krajorama) [inline model/textparse/protobufparse.go:419]: Nit: I see we are already using the first item to decide the type, which I think is assuming something that protobuf does not guarantee, but it's used in the code above as well so not a new problem. Maybe move to its own inline function?

**Review by krajorama — APPROVED** [review]
LGTM, tested overnight. Disclaimer: I'm not super familiar with the code, but makes sense to reset the state between reading classic histograms from native histograms.

File: model/textparse/protobufparse.go (hunk 1 above)
Comment (beorn7) [reply model/textparse/protobufparse.go:419]: In line 356, we skip empty metric families. So we can never reach the states `EntryHistogram` or `EntrySeries` if there is not at least one metric.

**Pull request merged**